cdm/gcdm/src/main/java/ucar/gcdm/GcdmConverter.java

```diff
@@ -0,0 +1,733 @@
+        return array;
+      }
+      case INT: {
+        int i = 0;
+        final int[] array = new int[data.getIntDataCount()];
+        for (int val : data.getIntDataList()) {
+          array[i++] = val;
+        }
+        return array;
+      }
+      case ENUM2:
+      case USHORT: {
+        int i = 0;
+        final short[] array = new short[data.getUintDataCount()];
+        for (int val : data.getUintDataList()) {
+          array[i++] = (short) val;
+        }
+        return array;
+      }
+      case ENUM4:
+      case UINT: {
+        int i = 0;
+        final int[] array = new int[data.getUintDataCount()];
+        for (int val : data.getUintDataList()) {
+          array[i++] = val;
+        }
+        return array;
+      }
+      case LONG: {
+        int i = 0;
+        final long[] array = new long[data.getLongDataCount()];
+        for (long val : data.getLongDataList()) {
+          array[i++] = val;
+        }
+        return array;
+      }
+      case ULONG: {
+        int i = 0;
+        final long[] array = new long[data.getUlongDataCount()];
+        for (long val : data.getUlongDataList()) {
+          array[i++] = val;
+        }
+        return array;
+      }
+      case FLOAT: {
+        int i = 0;
+        final float[] array = new float[data.getFloatDataCount()];
+        for (float val : data.getFloatDataList()) {
+          array[i++] = val;
+        }
+        return array;
+      }
+      case DOUBLE: {
+        int i = 0;
+        final double[] array = new double[data.getDoubleDataCount()];
+        for (double val : data.getDoubleDataList()) {
+          array[i++] = val;
+        }
+        return array;
+      }
+      case STRING: {
+        int i = 0;
+        final Object[] array = new Object[data.getStringDataCount()];
+        for (String val : data.getStringDataList()) {
+          array[i++] = val;
+        }
+        return array;
+      }
+      case OPAQUE: {
+        int i = 0;
+        final Object[] array = new Object[data.getByteDataCount()];
+        for (ByteString val : data.getByteDataList()) {
+          array[i++] = ByteBuffer.wrap(val.toByteArray());
+        }
+        return array;
+      }
+      default:
+        throw new IllegalStateException("Unknown datatype " + dataType);
+    }
+  }
+
+  private static Array decodeVlenData(Data vlenData) {
+    Preconditions.checkArgument(vlenData.getVlenDataCount() > 0);
+    final int[] shape = decodeShape(vlenData);
+    final int length = (int) Index.computeSize(shape);
+    Preconditions.checkArgument(length == vlenData.getVlenDataCount());
+    final Array[] storage = new Array[length];
+
+    for (int i = 0; i < length; i++) {
+      final Data inner = vlenData.getVlenData(i);
+      storage[i] = decodeData(inner);
+    }
+
+    return Array.makeVlenArray(shape, storage);
+  }
+
+  ////////////////////////////////////////////////////////////////
+
+  public static GcdmNetcdfProto.DataType convertDataType(DataType dataType) {
```

Comment:
I think this is fine for now, but there's gotta be a better way to do it... two hashmaps?

Reply:
Yes, that is definitely nicer! I used one bidirectional map in 81bd7991b6